src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -120,20 +123,47 @@ private static int native_strlen_byte(MemorySegment segment, long start) {
             if (start > 0) {
                 segment = segment.asSlice(start);
             }
-            long len = segment.byteSize() < MAX_TRIVIAL_SIZE
-                    ? (long)STRLEN_TRIVIAL.invokeExact(segment)
-                    : (long)STRLEN.invokeExact(segment);
+            long segmentSize = segment.byteSize();
+            final long len;
+            if (SIZE_T_IS_INT) {
+                if (segmentSize < MAX_TRIVIAL_SIZE) {
+                    len = (int)STRNLEN_TRIVIAL.invokeExact(segment, (int) segmentSize);
+                } else if (segmentSize < Integer.MAX_VALUE) {
```

Comment:
size_t is unsigned. So it should be fine to use the native variant as long as the upper 32 bits are not set.